src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/ValueTypeConverter.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Data.SqlTypes;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClient
+{
+    /// <summary>
+    /// Delegate representing a converter to pass in by reference.
+    /// </summary>
+    /// <typeparam name="TIn"></typeparam>
+    /// <typeparam name="TOut"></typeparam>
+    /// <param name="input"></param>
+    /// <returns></returns>
+    public delegate TOut RefConverter<TIn, TOut>(ref TIn input);
+
+    /// <summary>
+    /// Converts value types from a generic to the desired, underlying type without boxing.
+    /// </summary>
+    /// <typeparam name="TIn"></typeparam>
+    /// <typeparam name="TOut"></typeparam>
+    public static class ValueTypeConverter<TIn, TOut>
+    {
+        /// <summary>
+        /// Converts the value to the TOut type
+        /// </summary>
+        public static readonly RefConverter<TIn, TOut> Convert;
+
+        static ValueTypeConverter()
+        {
+            var paramExpr = Expression.Parameter(typeof(TIn).MakeByRefType(), "input");
+
+            var lambda = typeof(TIn) != typeof(TOut)
+                ? Expression.Lambda<RefConverter<TIn, TOut>>(Expression.Convert(paramExpr, typeof(TOut)), paramExpr)
+                : Expression.Lambda<RefConverter<TIn, TOut>>(paramExpr, paramExpr)
+            ;
+
+            Convert = lambda.Compile();
+        }
```

Comment:
This is the "magic" bit isn't it. It takes the type parameters and causes a delegate to be emitted with a correct cast implementation in it. 
I think this is the only place that reflection emit is used and I wonder if there's another way to approach this? Given that the type map is statis can it be approaches with static code in a similar way to GetFieldValueAsync+SqlBuffer ? I worry that some platforms may not permit reflection emit.

Reply:
Interesting, I didn’t consider this kind of code to be platform specific, but that makes sense. I can take a look to refactor this to be more similar to the example you referenced.

Reply:
So I took a look at the internals of `SqlDataReader.GetFieldValueAsync` and that looks to do some heaving boxing as well.

```csharp
internal T Int32As<T>()
{
        ThrowIfNull();
        return (T)(object)_value.
}
```

I would say this code is a candidate for using the `ValueTypeConverter` as well.  Maybe there is a combination of my approach and this approach to detect if `Reflection.Emit` is platform supported and conditionally create the delegate?

Reply:
Nope, there's trickery here 😁 Take a look at https://github.com/dotnet/corefx/pull/34999 and the explanation example I give. While that pattern looks like it'll box is won't.

Reply:
Oh that is AWESOME. You gotta love the power of the JIT sometimes. 

This is a prime example of just looking at the code not always being enough and actually benchmarking. 

Very very cool. 

I pulled my example from [here](https://stackoverflow.com/questions/3343551/how-to-cast-a-value-of-generic-type-t-to-double-without-boxing)

Reply:
In your `EnumerableDataReaderFactoryBuilder` if you add in two declarations:

```
private static readonly object BoxedTrue = (object)true;
private static readonly object BoxedFalse = (object)false;
``
and then in `Add` add this:
```
            if (typeof(TColumn) == typeof(bool))
            {
                objExpression = o => ((bool)(object)func(o) ? BoxedTrue : BoxedFalse);
            }
            else
            {
                objExpression = (o) => func(o);
            }
```

You eliminate the boxing of retrieved bools and return a pre-boxed reference to one of the new constants which reduces overhead from your test harness.

Reply:
While that is definitely an improvement, it will help reduce some overheard using the current `SqlBulkCopy`, vs the proposed which attempts to avoid boxing entirely.